db/migrate/20200501155054_installed_package_unique_nvrea.rb

```diff
@@ -0,0 +1,64 @@
+class InstalledPackageUniqueNvrea < ActiveRecord::Migration[5.2]
+  def fix_missing_attributes
+    #bug in dynflow may have resulted in old code running and not properly populating fields
+    # This block is basically copied from 20200129172534_add_epoch_version_release_arch_to_katello_installed_packages.rb
+    epoch_non_0 = ::Katello::Rpm.where.not(epoch: [0, nil]).pluck(:nvra, :epoch).to_h
+    installed_packages = []
+    ::Katello::InstalledPackage.where(:nvrea => nil).each do |pkg|
+      epoch = epoch_non_0[pkg.nvra] || "0"
+
+      attributes_hash = ::Katello::Util::Package.parse_nvrea(pkg.nvra)
+      attributes_hash[:epoch] = epoch
+      attributes_hash[:nvra] = pkg.nvra
+      if epoch == "0"
+        attributes_hash[:nvrea] = pkg.nvra
+      else
+        attributes_hash[:nvrea] = "#{pkg.name}-#{epoch}:#{attributes_hash[:version]}-"\
+                                  "#{attributes_hash[:release]}.#{attributes_hash[:arch]}"
+      end
+
+      installed_packages << ::Katello::InstalledPackage.new(attributes_hash)
+    end
+    ::Katello::InstalledPackage.import(installed_packages, validate: false, batch_size: 50_000,
+                                       on_duplicate_key_update: {conflict_target: [:nvra],
+                                                                 columns: [:nvrea, :epoch, :version, :release, :arch]})
+  end
+
+  def consolidate_duplicate_nvreas
```

Comment:
Why is this method necessary?  Are duplicate installed packages being created as part of the bug?

Reply:
In theory, it shouldn't happen, but anytime we take a column and make it unique we need to ensure there are no duplicates.  Otherwise users would be stuck in the middle of a db migration